### PR TITLE
[GOBBLIN-1752] Fix race condition where FSTemplateCatalog would update at the same t…

### PR DESCRIPTION
**Pull request by Will-Lo**
…ime as the FSFlowGraph if their files are updated at the same time as the flowgraph monitor

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1752

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

`FSFlowGraphMonitor` and the `ObservingFsFlowTemplateCatalog` have a race condition if they are updated at the same time, the FSFlowGraphMonitor will not reload itself and can be put into a bad state where there are jobs or edges that are incomplete due to deleting the cache mid-update.

Instead, we should have `FSFlowGraphMonitor` control the updates to both the flowgraph and the templateCatalog through a configuration `gobblin.service.fsFlowGraphMonitor.monitorTemplateChanges`, which assumes that the templateCatalog is colocated with the flowgraph in the same parent folder.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitorTest.java [hunk 1]
```diff
@@ -227,6 +234,22 @@ public void testSharedFlowgraphHelper() throws Exception {
   }
 
   @Test (dependsOnMethods = "testSharedFlowgraphHelper")
```
Comment (umustafi) [inline gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitorTest.java:236]: how does this test the new logic above? Is it testing the new lock used monitors changes?

Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitorTest.java:236]: It's testing that FSFlowGraphMonitor will reload the flowgraph even when there are only changes done to the flow templates, I added some comments to clarify this in the test.

Previously FSFlowGraphMonitor wouldn't listen to the templateCatalog, and would only concern itself on changes made to the flowgraph.

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/template_catalog/UpdatableFSFFlowTemplateCatalogTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    private FSFlowTemplateCatalog flowTemplateCatalog;
+    private String configKey;
+
+    @Override
+    public String apply(Void input) {
+      try {
+        return this.flowTemplateCatalog.getFlowTemplate(this.flowTemplateCatalogUri).getRawTemplateConfig().getString(this.configKey);
+      } catch (SpecNotFoundException | JobTemplate.TemplateException | IOException | URISyntaxException e) {
+        throw new RuntimeException(e);
+      }
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/template_catalog/UpdatableFSFFlowTemplateCatalogTest.java]: Removing this to simplify the tests, before it was used as part of a timing function but its unnecessary now.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitor.java [hunk 3]
```diff
@@ -61,24 +63,29 @@ public class FsFlowGraphMonitor extends AbstractIdleService implements FlowGraph
...
+  private Path flowGraphPath;
+  private Path observedPath;
   private final MultiHopFlowCompiler compiler;
   private final CountDownLatch initComplete;
   private static final Config DEFAULT_FALLBACK = ConfigFactory.parseMap(ImmutableMap.<String, Object>builder()
       .put(ConfigurationKeys.FLOWGRAPH_ABSOLUTE_DIR, DEFAULT_FS_FLOWGRAPH_MONITOR_ABSOLUTE_DIR)
       .put(ConfigurationKeys.FLOWGRAPH_BASE_DIR, DEFAULT_FS_FLOWGRAPH_MONITOR_FLOWGRAPH_DIR)
       .put(ConfigurationKeys.FLOWGRAPH_POLLING_INTERVAL, DEFAULT_FLOWGRAPH_POLLING_INTERVAL)
       .put(ConfigurationKeys.FLOWGRAPH_JAVA_PROPS_EXTENSIONS, ConfigurationKeys.DEFAULT_PROPERTIES_EXTENSIONS)
+      .put(MONITOR_TEMPLATE_CATALOG_CHANGES, false)
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitor.java:75]: why is this false by default? I thought we want this enabled to refresh from both together

**Review by umustafi — APPROVED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitor.java (hunk 3 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitor.java:75]: Going under the assumption that the flowgraph and the templates are not colocated in the same directory for other setups.

**Pull request merged**